### ISSUE TITLE
Events are Appearing Without Timezone Offset

### ISSUE THREAD

**Issue by mrjarbenne** [hwdsbcommons/tec-ical-importer#19]
Full day events are working fine, but events that have a time set are appearing without the timezone data (so in our case in EST, all events are 4 hours early). We dealt with something similar to this previously https://github.com/hwdsbcommons/tec-ical-importer/commit/7e8478ef0d0399f622f4bb312563d9467df7f524 but it seems to be rearing its head again. I've set up a couple of feeds on https://sandbox.commons.hwdsb.on.ca to test (https://sandbox.commons.hwdsb.on.ca/event/an-event-at-8am-on-wed/) and can share an outlook feed if necessary.

**Comment by r-a-y** [comment]
Do you mind if I discuss certain aspects about the iCal file being used on the Sandbox site?  Or should I ping you via email about them?

**Comment by mrjarbenne** [comment]
Discuss away. It’s a test feed.

**Comment by r-a-y** [comment]
For the followng:

```
SUMMARY:A Final Test Event
DTSTART;TZID=(UTC) Coordinated Universal Time:20180122T200000
DTEND;TZID=(UTC) Coordinated Universal Time:20180122T210000
CLASS:PUBLIC
PRIORITY:5
```

The time is using UTC.  So do you expect the time to be at 4pm when imported into WordPress (due to the timezone change), instead of 8pm?

**Comment by mrjarbenne** [comment]
Well that's a bit of a mess isn't it. The time zone switches back and forth in the ics file. 

Here's a snippet from further down, with a different TZID: 

``
BEGIN:VEVENT
DESCRIPTION:\n
UID:040000008200E00074C5B7101A82E008000000009D6D654F32CCD301000000000000000
 01000000067941390F2F67A43BE1CDA2B7759CF5B
SUMMARY:An Event at 8AM on Wed
DTSTART;TZID="(UTC-05:00) Eastern Time (US & Canada)":20180411T080000
DTEND;TZID="(UTC-05:00) Eastern Time (US & Canada)":20180411T083000
CLASS:PUBLIC
PRIORITY:5
DTSTAMP:20180406T165528Z
TRANSP:OPAQUE
STATUS:CONFIRMED
SEQUENCE:0
LOCATION:
X-MICROSOFT-CDO-APPT-SEQUENCE:0
X-MICROSOFT-CDO-BUSYSTATUS:BUSY
X-MICROSOFT-CDO-INTENDEDSTATUS:BUSY
X-MICROSOFT-CDO-ALLDAYEVENT:FALSE
X-MICROSOFT-CDO-IMPORTANCE:1
X-MICROSOFT-CDO-INSTTYPE:0
X-MICROSOFT-DONOTFORWARDMEETING:FALSE
X-MICROSOFT-DISALLOW-COUNTER:FALSE
END:VEVENT
``

**Comment by mrjarbenne** [comment]
I just added another event on the ical feed that's feeding the hub category on sandbox (I might delete the other one so we are only dealing with one ical file) on the 12th, at 8AM, which is appearing at 4AM.

In Settings, I've turned on the Time Zone indicator, so that the time zone appears after the time. Showing as EDT in the screenshot

![2018-04-06_13-05-05](https://user-images.githubusercontent.com/480667/38434262-c18a5af4-399b-11e8-9c3f-e4d318ea0056.png)


![2018-04-06_13-03-38](https://user-images.githubusercontent.com/480667/38434179-6f8089f4-399b-11e8-83c3-a57527577c0f.png)

In Outlook, I've confirmed that the calendar we are feeding from is indeed set to the right time zone: 

![2018-04-06_13-07-03](https://user-images.githubusercontent.com/480667/38434195-7f3c2966-399b-11e8-8e95-54c26b52bdf5.png)

**Comment by r-a-y** [comment]
Okay, I've added some fixes on production and have manually resynced the "hub" iCalendar.

I believe I've fixed the issues.  Can you try a bunch of other iCals for sanity's sake?  Once we're in the clear, I'll commit the fix.

**Comment by mrjarbenne** [comment]
Ok. I've tested a few different calendars from both Outlook and Google and I think you have this solved. Thanks for sorting it out.

**Comment by mrjarbenne** [comment]
Well, it looks like they changed something again and broke this 👎 . Events set to 8AM are appearing as 4AM, although the Events Calendar seems to be able to identify that the events are from the correct time zone, even though the time offset is wrong.

![2018-06-14_22-33-49](https://user-images.githubusercontent.com/480667/41447515-a0a17070-7023-11e8-86bf-8e61f4c58f65.gif)

**Comment by r-a-y** [comment]
Yeah, it looks like a change in TEC caused this.

Looking into it.

**Comment by r-a-y** [comment]
So, I have a potential fix for this.

In the past, we had to manually calculate the date for each event using the GMT offset of the current site's timezone:
https://github.com/hwdsbcommons/tec-ical-importer/blob/master/includes/tec-ical-parser.php#L169

It appears TEC might have made some improvements so we no longer need to do this.  I've tested locally and this works with the sample Microsoft iCal from the GIF.  However, this might potentially break other calendars that are not Microsoft iCals.

Can you do a quick test with Google Calendar iCals?  Fix is on production.  For the Microsoft iCals, you'll need to remove the event and manually resync.

**Comment by mrjarbenne** [comment]
I tested it over here: https://sandbox.commons.hwdsb.on.ca/events/

If you look at the three events in this week (20,21,22), you are correct that the fix works for Microsoft (the event on the 21st), but for Google iCal (22nd), and the ics feed off of D2L (20th), the events are landing 4 hours past the scheduled time, so the offset is 4 hours the other direction.

**Comment by r-a-y** [comment]
So it looks like the problem was with Outlook's iCals and with the SG-iCalendar library.

I've patched up SG-iCalendar to address this.  I tested all three iCal types (D2L, Outlook and Google) and the time appears to be correct now.

Can you confirm?

**Comment by mrjarbenne** [comment]
Yes. I can confirm it's all looking correct now. I think you can commit that change. Thanks for sorting that out.

**Comment by mrjarbenne** [comment]
@r-a-y  I'm re-opening this (again). Events are now syncing from Outlook 1 hour later than they should be (an event posted in Outlook at 8:00 is appearing at 9). I'm investigating when this may have begun happening. Just reported today.

**Comment by r-a-y** [comment]
Is the example on the Sandbox site?  If so, let me know which iCal and I can grab the ICS link.  If not, pass the link to me privately via email so I can take a look.

**Comment by r-a-y** [comment]
So the problem was due to some prior fixes to address other timezone strings used by Outlook.

The first one tried to fix timezones with the following format, "(UTC-05:00) Eastern Time (US & Canada)".  The second one was for the following type of timezone, "Eastern Standard Time".

The second fix was wrong as it mapped "Eastern Standard Time" to "America/Cancun", which is totally different from "America/New_York" or "America/Toronto".

So in order for both timezone formats to work, I needed to do a better check to map the correct timezone format.  Should be finally fixed now, however you will need to delete and resync these events for sites using Outlook iCals.
